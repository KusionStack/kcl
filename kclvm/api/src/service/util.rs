use crate::gpyrpc::ExecProgramArgs;

/// Transform the str with zero value into [`Option<String>`]
#[inline]
pub(crate) fn transform_str_para(para: &str) -> Option<String> {
    if para.is_empty() {
        None
    } else {
        Some(para.to_string())
    }
}

#[inline]
pub(crate) fn transform_exec_para(
    exec_args: &Option<ExecProgramArgs>,
) -> anyhow::Result<kclvm_runner::ExecProgramArgs> {
    Ok(match exec_args {
        Some(exec_args) => {
            let args_json = serde_json::to_string(exec_args)?;
            kclvm_runner::ExecProgramArgs::from_str(args_json.as_str())
        }
        None => kclvm_runner::ExecProgramArgs::default(),
    })
}
