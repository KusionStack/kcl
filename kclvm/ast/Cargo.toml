[package]
name = "kclvm-ast"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
rustc_span = { path = "../3rdparty/rustc_span" }
serde = { version = "1", features = ["derive"] }
serde_json = "1.0"
indexmap = "1.0"
fancy-regex = "0.7.1"

kclvm-span = {path = "../span", version = "0.1.0"}
kclvm-error = {path = "../error", version = "0.1.0"}

[dev-dependencies]
kclvm-parser = {path = "../parser", version = "0.1.0"}
pretty_assertions = "1.2.1"
