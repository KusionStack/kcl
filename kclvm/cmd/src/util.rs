use clap::ArgMatches;

#[inline]
pub(crate) fn strings_from_matches(matches: &ArgMatches, key: &str) -> Option<Vec<String>> {
    matches.values_of(key).map(|files| {
        files
            .into_iter()
            .map(|v| v.to_string())
            .collect::<Vec<String>>()
    })
}

#[inline]
pub(crate) fn string_from_matches(matches: &ArgMatches, key: &str) -> Option<String> {
    matches.value_of(key).map(|v| v.to_string())
}

#[inline]
pub(crate) fn bool_from_matches(matches: &ArgMatches, key: &str) -> Option<bool> {
    let occurrences = matches.occurrences_of(key);
    if occurrences > 0 {
        Some(true)
    } else {
        None
    }
}

#[inline]
pub(crate) fn u32_from_matches(matches: &ArgMatches, key: &str) -> Option<u32> {
    let occurrences = matches.occurrences_of(key);
    if occurrences > 0 {
        Some(occurrences as u32)
    } else {
        None
    }
}
