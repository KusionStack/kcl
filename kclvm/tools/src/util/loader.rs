use std::fs;

use anyhow::{bail, Context, Result};

pub(crate) trait Loader<T> {
    fn load_ast(&self) -> Result<T>;
}

pub(crate) struct LoaderInner {
    content: String,
}

pub(crate) enum LoaderKind {
    YAML,
    JSON,
}

impl LoaderInner {
    fn new_with_file_path(file_path: &str) -> Result<Self> {
        let content = fs::read_to_string(file_path)
            .with_context(|| format!("Failed to Load '{}'", file_path))?;

        Ok(Self { content })
    }

    fn new_with_str(s: &str) -> Result<Self> {
        Ok(Self {
            content: s.to_string(),
        })
    }

    fn get_content(&self) -> &str {
        &self.content
    }
}

pub(crate) struct DataLoader {
    kind: LoaderKind,
    inner: LoaderInner,
}

impl DataLoader {
    pub(crate) fn new_with_file_path(loader_kind: LoaderKind, file_path: &str) -> Result<Self> {
        Ok(Self {
            kind: loader_kind,
            inner: LoaderInner::new_with_file_path(file_path)
                .with_context(|| format!("Failed to Load '{}'", file_path))?,
        })
    }

    pub(crate) fn new_with_str(loader_kind: LoaderKind, s: &str) -> Result<Self> {
        Ok(Self {
            kind: loader_kind,
            inner: LoaderInner::new_with_str(s)
                .with_context(|| format!("Failed to Load '{}'", s))?,
        })
    }

    pub(crate) fn get_data(&self) -> &str {
        &self.inner.get_content()
    }
}

impl Loader<serde_json::Value> for DataLoader {
    fn load_ast(&self) -> Result<serde_json::Value> {
        let v = match self.kind {
            LoaderKind::JSON => serde_json::to_value(&self.get_data())
                .with_context(|| format!("Failed to String '{}' to Json", self.get_data()))?,
            _ => {
                bail!("Failed to String to Json Value")
            }
        };

        Ok(v)
    }
}

impl Loader<serde_yaml::Value> for DataLoader {
    fn load_ast(&self) -> Result<serde_yaml::Value> {
        let v = match self.kind {
            LoaderKind::YAML => serde_yaml::to_value(&self.get_data())
                .with_context(|| format!("Failed to String '{}' to Yaml", self.get_data()))?,
            _ => {
                bail!("Failed to String to Yaml Value")
            }
        };

        Ok(v)
    }
}
