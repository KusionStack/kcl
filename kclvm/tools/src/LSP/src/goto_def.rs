//! GotoDefinition for KCL
//! Github Issue: https://github.com/kcl-lang/kcl/issues/476
//! Now supports goto definition for the following situation:
//! + variable
//! + schema definition
//! + mixin definition
//! + schema attr
//! + attr type

use indexmap::{IndexMap, IndexSet};
use kclvm_ast::pos::{ContainsPos, GetPos};

use kclvm_ast::ast::{Expr, Identifier, ImportStmt, Node, Program, Stmt};
use kclvm_compiler::pkgpath_without_prefix;
use kclvm_error::Position as KCLPos;

use kclvm_sema::resolver::scope::{ProgramScope, Scope, ScopeObject};
use kclvm_sema::ty::{DictType, SchemaType};
use lsp_types::{GotoDefinitionResponse, Url};
use lsp_types::{Location, Range};
use std::cell::RefCell;
use std::path::Path;
use std::rc::Rc;

use crate::to_lsp::lsp_pos;
use crate::util::{
    fix_missing_identifier, get_pkg_scope, get_pos_from_real_path, get_real_path_from_external,
    inner_most_expr_in_stmt,
};

// Navigates to the definition of an identifier.
pub(crate) fn goto_definition(
    program: &Program,
    kcl_pos: &KCLPos,
    prog_scope: &ProgramScope,
) -> Option<lsp_types::GotoDefinitionResponse> {
    match program.pos_to_stmt(kcl_pos) {
        Some(node) => match node.node {
            Stmt::Import(stmt) => goto_def_for_import(&stmt, kcl_pos, prog_scope, program),
            _ => match find_def(node.clone(), kcl_pos, prog_scope) {
                Some(def) => positions_to_goto_def_resp(&def.get_positions()),
                None => None,
            },
        },
        None => None,
    }
}

// Todo: fix ConfigExpr
// ```kcl
// schema Person:
//     name: str
//     data: Data

// schema Data:
//     id: int

// person = Person {
//     data.id = 1
//     data: {
//         id = 1
//     }
//     data: Data {
//         id = 3
//     }
// }

#[derive(Debug)]
pub enum Definition {
    Object(ScopeObject),
    Scope(Scope),
}

impl Definition {
    pub(crate) fn get_positions(&self) -> IndexSet<(KCLPos, KCLPos)> {
        let mut positions = IndexSet::new();
        match self {
            Definition::Object(obj) => {
                positions.insert((obj.start.clone(), obj.end.clone()));
            }
            Definition::Scope(scope) => match &scope.kind {
                kclvm_sema::resolver::scope::ScopeKind::Package(filenames) => {
                    for file in filenames {
                        let dummy_pos = KCLPos {
                            filename: file.clone(),
                            line: 1,
                            column: None,
                        };
                        positions.insert((dummy_pos.clone(), dummy_pos));
                    }
                }
                _ => {
                    positions.insert((scope.start.clone(), scope.end.clone()));
                }
            },
        }
        positions
    }
}

pub(crate) fn find_def(
    node: Node<Stmt>,
    kcl_pos: &KCLPos,
    prog_scope: &ProgramScope,
) -> Option<Definition> {
    fn pre_process_identifier(id: Node<Identifier>, pos: &KCLPos) -> Identifier {
        if !id.contains_pos(pos) && id.node.names.is_empty() {
            return id.node.clone();
        }

        let mut id = id.node.clone();
        let mut names = vec![];
        for name in id.names {
            names.push(name.clone());
            if name.contains_pos(pos) {
                break;
            }
        }
        id.names = fix_missing_identifier(&names);
        if !id.pkgpath.is_empty() {
            id.names[0].node = pkgpath_without_prefix!(id.pkgpath);
        }
        id
    }

    let (inner_expr, parent) = inner_most_expr_in_stmt(&node.node, kcl_pos, None);
    if let Some(expr) = inner_expr {
        if let Expr::Identifier(id) = expr.node {
            let id_node = Node::node_with_pos(
                id.clone(),
                (
                    expr.filename,
                    expr.line,
                    expr.column,
                    expr.end_line,
                    expr.end_column,
                ),
            );
            let id = pre_process_identifier(id_node, kcl_pos);
            match parent {
                Some(schema_expr) => {
                    if let Expr::Schema(schema_expr) = schema_expr.node {
                        let schema_def =
                            find_def(node, &schema_expr.name.get_end_pos(), prog_scope);
                        if let Some(schema) = schema_def {
                            match schema {
                                Definition::Object(obj) => {
                                    let schema_type = obj.ty.into_schema_type();
                                    return find_attr_in_schema(
                                        &schema_type,
                                        &id.names,
                                        &prog_scope.scope_map,
                                    );
                                }
                                Definition::Scope(_) => {
                                    //todo
                                }
                            }
                        }
                    }
                }
                None => {
                    if let Some(inner_most_scope) = prog_scope.inner_most_scope(&kcl_pos) {
                        return resolve_var(&id.names, &inner_most_scope, &prog_scope.scope_map);
                    }
                }
            }
        }
    }
    None
}

/// Similar to vars.rs/resolver_var, find a ScopeObj corresponding to the definition of identifier
pub(crate) fn resolve_var(
    node_names: &[Node<String>],
    current_scope: &Scope,
    scope_map: &IndexMap<String, Rc<RefCell<Scope>>>,
) -> Option<Definition> {
    let names = node_names
        .iter()
        .map(|node| node.node.clone())
        .collect::<Vec<String>>();
    match names.len() {
        0 => None,
        1 => {
            let name = names[0].clone();
            match current_scope.lookup(&name) {
                Some(obj) => match obj.borrow().kind {
                    kclvm_sema::resolver::scope::ScopeObjectKind::Module(_) => {
                        match scope_map.get(&name) {
                            Some(scope) => Some(Definition::Scope(scope.borrow().clone())),
                            None => None,
                        }
                    }
                    _ => Some(Definition::Object(obj.borrow().clone())),
                },
                None => None,
            }
        }
        _ => {
            let name = names[0].clone();
            match current_scope.lookup(&name) {
                Some(obj) => {
                    match &obj.borrow().ty.kind {
                        kclvm_sema::ty::TypeKind::Schema(schema_type) => {
                            find_attr_in_schema(schema_type, &node_names[1..], scope_map)
                        }
                        kclvm_sema::ty::TypeKind::Module(module_ty) => {
                            match scope_map.get(&pkgpath_without_prefix!(module_ty.pkgpath)) {
                                Some(scope) => {
                                    return resolve_var(
                                        &node_names[1..],
                                        &scope.borrow(),
                                        scope_map,
                                    );
                                }
                                None => None,
                            }
                        }
                        kclvm_sema::ty::TypeKind::Dict(DictType { attrs: _, .. }) => {
                            // Todo: find key def in dict
                            None
                        }
                        _ => None,
                    }
                }
                None => None,
            }
        }
    }
}

pub fn find_attr_in_schema(
    schema_type: &SchemaType,
    names: &[Node<String>],
    scope_map: &IndexMap<String, Rc<RefCell<Scope>>>,
) -> Option<Definition> {
    let schema_pkg_scope = get_pkg_scope(&schema_type.pkgpath, scope_map);
    let names = if schema_type.pkgpath.is_empty() {
        &names[1..]
    } else {
        names
    };
    for child in &schema_pkg_scope.children {
        let child_scope = child.borrow();
        if let kclvm_sema::resolver::scope::ScopeKind::Schema(schema_name) = &child_scope.kind {
            if schema_name == &schema_type.name {
                return resolve_var(&names, &child_scope, scope_map);
            }
        }
    }
    None
}

// Convert kcl position to GotoDefinitionResponse. This function will convert to
// None, Scalar or Array according to the number of positions
fn positions_to_goto_def_resp(
    positions: &IndexSet<(KCLPos, KCLPos)>,
) -> Option<GotoDefinitionResponse> {
    match positions.len() {
        0 => None,
        1 => {
            let (start, end) = positions.iter().next().unwrap().clone();
            Some(lsp_types::GotoDefinitionResponse::Scalar(Location {
                uri: Url::from_file_path(start.filename.clone()).unwrap(),
                range: Range {
                    start: lsp_pos(&start),
                    end: lsp_pos(&end),
                },
            }))
        }
        _ => {
            let mut res = vec![];
            for (start, end) in positions {
                res.push(Location {
                    uri: Url::from_file_path(start.filename.clone()).unwrap(),
                    range: Range {
                        start: lsp_pos(start),
                        end: lsp_pos(end),
                    },
                })
            }
            Some(lsp_types::GotoDefinitionResponse::Array(res))
        }
    }
}

fn goto_def_for_import(
    stmt: &ImportStmt,
    _kcl_pos: &KCLPos,
    _prog_scope: &ProgramScope,
    program: &Program,
) -> Option<GotoDefinitionResponse> {
    let pkgpath = &stmt.path;
    let mut real_path =
        Path::new(&program.root).join(pkgpath.replace('.', &std::path::MAIN_SEPARATOR.to_string()));
    let mut positions = get_pos_from_real_path(&real_path);

    if positions.is_empty() && !real_path.exists() {
        real_path =
            get_real_path_from_external(&stmt.pkg_name, pkgpath, program.root.clone().into());
    }

    positions = get_pos_from_real_path(&real_path);

    positions_to_goto_def_resp(&positions)
}

pub(crate) fn get_identifier_last_name(id: &Identifier) -> String {
    match id.names.len() {
        0 => "".to_string(),
        1 => id.names[0].node.clone(),
        _ => {
            if id.names.last().unwrap().node == *"" {
                // MissingExpr
                id.names.get(id.names.len() - 2).unwrap().node.clone()
            } else {
                id.names.last().unwrap().node.clone()
            }
        }
    }
}
