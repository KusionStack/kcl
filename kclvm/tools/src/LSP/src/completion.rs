//! Complete for KCL
//! Github Issue: https://github.com/kcl-lang/kcl/issues/476
//! Now supports code completion in treigger mode (triggered when user enters `.`),
//! and the content of the completion includes:
//!  + import path
//!  + schema attr
//!  + builtin function(str function)
//!  + defitions in pkg
//!  + system module functions

use std::io;
use std::{fs, path::Path};

use indexmap::IndexSet;
use kclvm_ast::ast::{Expr, ImportStmt, Node, Program, Stmt};
use kclvm_ast::pos::GetPos;
use kclvm_config::modfile::KCL_FILE_EXTENSION;

use kclvm_error::Position as KCLPos;
use kclvm_sema::builtin::{
    get_system_module_members, STANDARD_SYSTEM_MODULES, STRING_MEMBER_FUNCTIONS,
};
use kclvm_sema::resolver::scope::{ProgramScope, ScopeObjectKind};
use lsp_types::CompletionItem;

use crate::goto_def::{find_def, get_identifier_last_name, Definition};
use crate::util::{inner_most_expr_in_stmt, is_in_schema};

/// Computes completions at the given position.
pub(crate) fn completion(
    trigger_character: Option<char>,
    program: &Program,
    pos: &KCLPos,
    prog_scope: &ProgramScope,
) -> Option<lsp_types::CompletionResponse> {
    if let Some('.') = trigger_character {
        completion_dot(program, pos, prog_scope)
    } else {
        let mut completions: IndexSet<KCLCompletionItem> = IndexSet::new();

        completions.extend(completion_variable(pos, prog_scope));

        completions.extend(completion_attr(program, pos, prog_scope));

        Some(into_completion_items(&completions).into())
    }
}

/// Abstraction of CompletionItem in KCL
#[derive(Debug, Clone, PartialEq, Hash, Eq)]
pub(crate) struct KCLCompletionItem {
    pub label: String,
}

fn completion_dot(
    program: &Program,
    pos: &KCLPos,
    prog_scope: &ProgramScope,
) -> Option<lsp_types::CompletionResponse> {
    // Get the position of trigger_character '.'
    let pos = &KCLPos {
        filename: pos.filename.clone(),
        line: pos.line,
        column: pos.column.map(|c| c - 1),
    };

    match program.pos_to_stmt(pos) {
        Some(node) => match node.node {
            Stmt::Import(stmt) => completion_for_import(&stmt, pos, prog_scope, program),
            _ => Some(into_completion_items(&get_completion(node, pos, prog_scope)).into()),
        },
        None => None,
    }
}

/// Complete schema attr
fn completion_attr(
    program: &Program,
    pos: &KCLPos,
    prog_scope: &ProgramScope,
) -> IndexSet<KCLCompletionItem> {
    let mut completions: IndexSet<KCLCompletionItem> = IndexSet::new();

    if let Some((node, schema_expr)) = is_in_schema(program, pos) {
        let schema_def = find_def(node, &schema_expr.name.get_end_pos(), prog_scope);
        if let Some(schema) = schema_def {
            if let Definition::Object(obj) = schema {
                let schema_type = obj.ty.into_schema_type();
                completions.extend(schema_type.attrs.keys().map(|attr| KCLCompletionItem {
                    label: attr.clone(),
                }));
            }
        }
    }
    completions
}

/// Complete all usable scope obj in inner_most_scope
fn completion_variable(pos: &KCLPos, prog_scope: &ProgramScope) -> IndexSet<KCLCompletionItem> {
    let mut completions: IndexSet<KCLCompletionItem> = IndexSet::new();
    if let Some(inner_most_scope) = prog_scope.inner_most_scope(pos) {
        for (name, obj) in inner_most_scope.all_usable_objects() {
            match &obj.borrow().kind {
                kclvm_sema::resolver::scope::ScopeObjectKind::Module(module) => {
                    completions.insert(KCLCompletionItem {
                        label: module.name.clone(),
                    });
                }
                _ => {
                    completions.insert(KCLCompletionItem { label: name });
                }
            }
        }
    }
    completions
}

fn completion_for_import(
    stmt: &ImportStmt,
    _pos: &KCLPos,
    _prog_scope: &ProgramScope,
    program: &Program,
) -> Option<lsp_types::CompletionResponse> {
    let mut items: IndexSet<KCLCompletionItem> = IndexSet::new();
    let pkgpath = &stmt.path;
    let real_path =
        Path::new(&program.root).join(pkgpath.replace('.', &std::path::MAIN_SEPARATOR.to_string()));
    if real_path.is_dir() {
        if let Ok(entries) = fs::read_dir(real_path) {
            let mut entries = entries
                .map(|res| res.map(|e| e.path()))
                .collect::<Result<Vec<_>, io::Error>>()
                .unwrap();
            entries.sort();
            for path in entries {
                let filename = path.file_name().unwrap().to_str().unwrap().to_string();
                if path.is_dir() {
                    items.insert(KCLCompletionItem { label: filename });
                } else if path.is_file() {
                    if let Some(extension) = path.extension() {
                        if extension == KCL_FILE_EXTENSION {
                            items.insert(KCLCompletionItem {
                                label: path
                                    .with_extension("")
                                    .file_name()
                                    .unwrap()
                                    .to_str()
                                    .unwrap()
                                    .to_string(),
                            });
                        }
                    }
                }
            }
        }
    }
    Some(into_completion_items(&items).into())
}

pub(crate) fn get_completion(
    stmt: Node<Stmt>,
    pos: &KCLPos,
    prog_scope: &ProgramScope,
) -> IndexSet<KCLCompletionItem> {
    let (expr, parent) = inner_most_expr_in_stmt(&stmt.node, pos, None);
    match expr {
        Some(node) => {
            let mut items: IndexSet<KCLCompletionItem> = IndexSet::new();
            match node.node {
                Expr::Identifier(id) => {
                    let name = get_identifier_last_name(&id);
                    if !id.pkgpath.is_empty() && STANDARD_SYSTEM_MODULES.contains(&name.as_str()) {
                        items.extend(
                            get_system_module_members(name.as_str())
                                .iter()
                                .map(|s| KCLCompletionItem {
                                    label: s.to_string(),
                                })
                                .collect::<IndexSet<KCLCompletionItem>>(),
                        )
                    }

                    let def = find_def(stmt, pos, prog_scope);

                    if let Some(def) = def {
                        match def {
                            crate::goto_def::Definition::Object(obj) => {
                                match &obj.ty.kind {
                                    // builtin (str) functions
                                    kclvm_sema::ty::TypeKind::Str => {
                                        let binding = STRING_MEMBER_FUNCTIONS;
                                        for k in binding.keys() {
                                            items.insert(KCLCompletionItem {
                                                label: format!("{}{}", k, "()"),
                                            });
                                        }
                                    }
                                    // schema attrs
                                    kclvm_sema::ty::TypeKind::Schema(schema) => {
                                        for k in schema.attrs.keys() {
                                            if k != "__settings__" {
                                                items
                                                    .insert(KCLCompletionItem { label: k.clone() });
                                            }
                                        }
                                    }
                                    _ => {}
                                }
                            }
                            crate::goto_def::Definition::Scope(s) => {
                                for (name, obj) in &s.elems {
                                    if let ScopeObjectKind::Module(_) = obj.borrow().kind {
                                        continue;
                                    } else {
                                        items.insert(KCLCompletionItem {
                                            label: name.clone(),
                                        });
                                    }
                                }
                            }
                        }
                    }
                }
                Expr::Selector(select_expr) => {
                    let res = get_completion(stmt, &select_expr.value.get_end_pos(), prog_scope);
                    items.extend(res);
                }
                Expr::StringLit(_) => {
                    let binding = STRING_MEMBER_FUNCTIONS;
                    for k in binding.keys() {
                        items.insert(KCLCompletionItem {
                            label: format!("{}{}", k, "()"),
                        });
                    }
                }
                Expr::Config(_) => match parent {
                    Some(schema_expr) => {
                        if let Expr::Schema(schema_expr) = schema_expr.node {
                            let schema_def =
                                find_def(stmt, &schema_expr.name.get_end_pos(), prog_scope);
                            if let Some(schema) = schema_def {
                                match schema {
                                    Definition::Object(obj) => {
                                        let schema_type = obj.ty.into_schema_type();
                                        items.extend(
                                            schema_type
                                                .attrs
                                                .keys()
                                                .map(|s| KCLCompletionItem {
                                                    label: s.to_string(),
                                                })
                                                .collect::<IndexSet<KCLCompletionItem>>(),
                                        );
                                    }
                                    Definition::Scope(_) => {}
                                }
                            }
                        }
                    }
                    None => {}
                },
                _ => {}
            }

            items
        }
        None => IndexSet::new(),
    }
}

pub(crate) fn into_completion_items(items: &IndexSet<KCLCompletionItem>) -> Vec<CompletionItem> {
    items
        .iter()
        .map(|item| CompletionItem {
            label: item.label.clone(),
            ..Default::default()
        })
        .collect()
}
