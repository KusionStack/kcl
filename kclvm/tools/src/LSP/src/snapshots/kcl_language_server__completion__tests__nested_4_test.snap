---
source: tools/src/LSP/src/completion.rs
expression: "format!(\"{:?}\", labels)"
---
["ab"]
